### PR TITLE
Berty do reszty tasków

### LINKED ISSUE #7: Berty do reszty tasków

### PR DESCRIPTION
**Pull request by Ruruthia**

### CODE REVIEW

File: scripts/train_bert.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        gpus=config["gpus"],
+        max_epochs=config["max_epochs"],
+        callbacks=[checkpoint_callback],
+    )
+
+    trainer.fit(model, dataloaders["train"], dataloaders["val"])
+    wandb.finish()
+
+
+if __name__ == "__main__":
```
Comment (Sahcim) [inline scripts/train_bert.py:47]: dodaj argparser taki jak w `download_and_prepare_data.py`

File: thegreatknowledgeheist/models/bert.py [hunk 2]
```diff
@@ -1,24 +1,42 @@
...
+from transformers import (
+    BertForMultipleChoice,
+    BertForSequenceClassification,
+    BertForTokenClassification,
+)
 
 
-class SentimentBert(pl.LightningModule):
-    def __init__(self, config):
+class Bert(pl.LightningModule):
```
Comment (Sahcim) [inline thegreatknowledgeheist/models/bert.py]: Implementacja działa, ale rozdzieliłbym to w taki sposób: Mamy klase BaseBert i klasy na poszczególne taski które będą dziedziczyć po Basebert

File: thegreatknowledgeheist/models/bert.py [hunk 3]
```diff
@@ -1,24 +1,42 @@
...
+        elif self.task == "swag":
+            self.model = BertForMultipleChoice.from_pretrained(
+                "bert-base-uncased", num_labels=4
+            )
+        else:
+            raise NotImplemented(
+                "Possible tasks are amazon_polarity, acronym_identification and swag"
+            )
 
     def configure_optimizers(self):
```
Comment (Sahcim) [inline thegreatknowledgeheist/models/bert.py:25]: Dodać learning rate scheduler i weight decay ( bardzo proste a przydatne )

File: scripts/download_and_prepare_dataset.py [hunk 4]
```diff
@@ -27,6 +27,24 @@ def amazon_prepare_and_save(
...
+
+        # Add the same label to the new list of labels `n_subwords` times
+        labels.extend([label] * n_subwords)
+    row["labels"] = labels[: tokenizer.model_max_length] + [
+        0 for _ in range(tokenizer.model_max_length - len(labels))
+    ]
+    return row
+
+
 def acronyms_prepare_and_save(
```
Comment (Sahcim) [inline scripts/download_and_prepare_dataset.py]: Sprawdzić, dlaczego po zmianach używany jest tylko 1 worker

File: scripts/download_and_prepare_dataset.py [hunk 5]
```diff
@@ -27,6 +27,24 @@ def amazon_prepare_and_save(
     encoded_dataset.save_to_disk(os.path.join(save_path, split))
 
 
+def tokenize_and_preserve_labels(
```
Comment (Sahcim) [inline scripts/download_and_prepare_dataset.py]: wrzucić do środka funkcji `acronyms_prepare_and_save` zadna inna funkcja nie bedzie z tego korzystać

File: thegreatknowledgeheist/models/bert.py [hunk 6]
```diff
@@ -28,7 +46,15 @@ def forward(self, **inputs):
...
+            preds = torch.argmax(logits, dim=1)
+        elif self.task == "acronym_identification":
+            preds = logits.argmax(-1)
+        else:
+            raise NotImplemented(
+                "Possible tasks are amazon_polarity, acronym_identification and swag"
+            )
+
         correct_preds = torch.sum(preds == batch["labels"])
         self.log("train_loss", loss, on_step=False, on_epoch=True)
```
Comment (Sahcim) [inline thegreatknowledgeheist/models/bert.py]: niech logowanie w treningu będzie co step, łatwiej wtedy patrzeć na wykres

File: thegreatknowledgeheist/models/bert.py [hunk 7]
```diff
@@ -39,7 +65,14 @@ def training_step(self, batch, batch_idx):
     def validation_step(self, batch, batch_idx):
```
Comment (Sahcim) [inline thegreatknowledgeheist/models/bert.py:41]: pokazuje się accuracy większe niż 100 :(

File: thegreatknowledgeheist/models/bert.py (hunk 2 above)
Comment (Sahcim) [inline thegreatknowledgeheist/models/bert.py]: Dodać parametr który zamraża x ostatnich warstw ( tak jak w publikacji ). Bardzo przyspiesza to trening bo mogę zrobić 4x większy batch, a według publikacji nie stracimy przy tym accuracy

**Review by Sahcim — CHANGES_REQUESTED** [review]

File: scripts/train_bert.py (hunk 1 above)
Comment (Ruruthia) [reply scripts/train_bert.py:47]: Zrobione!

File: thegreatknowledgeheist/models/bert.py (hunk 2 above)
Comment (Ruruthia) [reply thegreatknowledgeheist/models/bert.py]: Zrobione!

File: scripts/download_and_prepare_dataset.py (hunk 5 above)
Comment (Ruruthia) [reply scripts/download_and_prepare_dataset.py]: Zrobione!

File: thegreatknowledgeheist/models/bert.py (hunk 6 above)
Comment (Ruruthia) [reply thegreatknowledgeheist/models/bert.py]: Zrobione!

File: thegreatknowledgeheist/models/bert.py (hunk 7 above)
Comment (Ruruthia) [reply thegreatknowledgeheist/models/bert.py:41]: Chyba zrobione xD

File: scripts/train_bert.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    parser = ArgumentParser()
+    parser.add_argument("config_path", type=str, help="Path to config yaml")
+    args = parser.parse_args()
+    config = load_yaml(args.config_path)
+    dataloaders = get_dataloaders(
+        dataset_name=config["task"],
+        path_to_dataset=f"{config['dataset_path']}/{config['task']}",
```
Comment (AdrianUrbanski) [inline scripts/train_bert.py:54]: Zmieńmy to na path_to_dataset=config['dataset_path'], albo zmieńmy skrypt download_and_prepare_dataset żeby on zawsze zapisywał do podkatalogu z nazwą taska.

Obecnie żeby to zadziałało, to w download_and_prepare_dataset trzeba podać data/swag, a tutaj po prostu data. Więc przed użyciem skryptu do pobierania danych trzeba przeczytać kod tego skryptu, żeby wiedzieć jak nazwać podkatalogi do których pobieramy dane.

Comment (Sahcim) [reply scripts/train_bert.py:54]: poprawiam

**Review by Sahcim — APPROVED** [review]
